composition-js/src/merging/merge.ts

```diff
@@ -2058,6 +2072,69 @@ class Merger {
       });
     }
   }
+  
+  private removeOverriddenFieldFromFieldSet <T extends FieldDefinition<ObjectType | InterfaceType> | InputFieldDefinition>(
+    source: T, 
+    overriddenCoordinates: Set<string> | undefined, 
+    sourceMeta: FederationMetadata,
+  ) {
+    const providesDirective = sourceMeta.providesDirective();
+    const applications = source.appliedDirectivesOf(providesDirective);
+    assert(applications.length <= 1, () => `Found more than one application of ${providesDirective} on ${source}`);
+    if (applications.length === 0) {
+      return undefined;
+    }
+    const fields: string = applications[0].arguments().fields;
+    const parent = applications[0].parent;
+    const parentType = baseType(parent.type!);
+    
+    if (!overriddenCoordinates) {
+      return fields;
+    }
+
+    if (!isCompositeType(parentType)) {
+      return undefined;
+    }
+    
+    // when we parse the selection set, we will have a list of selections. Anything that eventually goes to a overridden field 
+    // cannot be considered valid in the supgraph
+    let validSelectionsIndex: boolean[] = [];
+    
+    const selectionSet = parseSelectionSet({
+      parentType,
+      source: fields,
+      fieldAccessor: (t, f) => {
+        const field = t.field(f);
+        
+        // Every time we get back to to seeing the parentType, we know that it is a new selection. 
+        // Assume it's valid until we see a contradiction
+        if (t.name === parentType.name) {
+          validSelectionsIndex.push(true);
+        } else if (isInterfaceType(parentType)) {
+          if (parentType.allImplementations().find(member => member.name === t.name)) {
+            validSelectionsIndex.push(true);
+          }
+        } else if (isUnionType(parentType)) {
+          if (parentType.members().find(member => t.name === member.type.name)) {
+            validSelectionsIndex.push(true);
+          }
+        }
+        assert(field, 'field should exist on type');
+        if (overriddenCoordinates.has(field.coordinate)) {
+          validSelectionsIndex[validSelectionsIndex.length-1] = false;
+        }
+        return field;
+      }
+    });
+    
+    // now we should have a SelectionSet with an array of _selections the same length as `validSelectionsIndex`
+    // We should be able to filter out the ones that are invalid and we'll be left with the valid selection string
+    const selections = selectionSet.selections();
+    assert(selections.length === validSelectionsIndex.length, 'selection parsing failed');
```

Comment:
would it be possible to have some `overriddenCoordinates` that are applicable to different `@provides`? (so there would be no field set filtering required for this `@provides`)